mantidproject/mantid
# HistogramData module

There is a [new `rst` file tries to explain some details](https://raw.githubusercontent.com/mantidproject/mantid/a9e3891a567cce5b1f164b4ed2bb43b1f121a3bc/docs/source/concepts/HistogramData.rst) ([github-rendered version that shows the drawings, but unfortunately not the code examples](https://github.com/mantidproject/mantid/blob/a9e3891a567cce5b1f164b4ed2bb43b1f121a3bc/docs/source/concepts/HistogramData.rst), build `docs-html` locally for best reading experience instead), it might be worth having a look at it before the review.

**Code review:**
This is a very large and critical change, I would like to suggest to split the review into several parts:

1a. Review everything in the new module `HistogramData`.
1b. Review the interface of `ISpectrum` (including `Histogram1D`/`EventList`) and `MatrixWorkspace`.
2. Feedback to me, I will adapt where necessary, and then review the full set of changes myself. (**done**)
3. Review of everything else (existing Mantid code).

I think we should have more than one reviewer for both 1) and 3).

**To test:**

This is a global change, there is no individual test. If you have access to any Python scripts using Mantid that are not part of our automated testing it would be good to check if they still work.

Fixes #16086.
Fixes #16096.
Fixes #16282.

Note there there will be several follow-up PRs once this has been merged, in particular,
- #16298 will store `X` uncertainties for points instead of bin edges (removing parts of the `Dx` interface)
- #16661 might add some (yet to be determined) degree of validation for data in histograms, and also adds generators for easier creation of histograms, e.g., `setBinEdges(1000, LinearGenerator(0, 0.1));`
- @LamarMoore has started refactoring to replace uses of the old `X`, `Y`, `E` interface methods by the new ones.



---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

This has been reviewed by several TSC members, namely @martyngigg and Owen Arnold. The changes are largely internal. This is well-structured, documented and tested.
